src/Test/ImplicitMethodsIntegrationTest.php

```diff
@@ -115,37 +173,68 @@ public function testExplicitRequest(string $method, MiddlewareInterface $middlew
-        $handler = new class ($finalHandler->reveal(), $middleware) implements RequestHandlerInterface
-        {
-            /** @var RequestHandlerInterface */
-            private $handler;
+        $request = new ServerRequest(['REQUEST_METHOD' => $method], [], '/api/v1/me', $method);
 
-            /** @var MiddlewareInterface */
-            private $middleware;
-
-            public function __construct(RequestHandlerInterface $handler, MiddlewareInterface $middleware)
-            {
-                $this->handler = $handler;
-                $this->middleware = $middleware;
-            }
-
-            public function handle(ServerRequestInterface $request) : ResponseInterface
-            {
-                return $this->middleware->process($request, $this->handler);
-            }
-        };
-
-        $request = new ServerRequest([], [], '/api/v1/me', $method);
-
-        $response = $routeMiddleware->process($request, $handler);
+        $response = $pipeline->process($request, $finalHandler->reveal());
 
         $this->assertEquals(StatusCode::STATUS_OK, $response->getStatusCode());
         $this->assertSame('FOO BAR BODY', (string) $response->getBody());
         $this->assertTrue($response->hasHeader('foo-bar'));
         $this->assertSame('baz', $response->getHeaderLine('foo-bar'));
     }
 
+    public function withoutImplicitMiddleware()
+    {
+        // request method, array of allowed methods for a route.
+        yield 'HEAD: get' => [RequestMethod::METHOD_HEAD, [RequestMethod::METHOD_GET]];
+        yield 'HEAD: post' => [RequestMethod::METHOD_HEAD, [RequestMethod::METHOD_POST]];
+        yield 'HEAD: get, post' => [RequestMethod::METHOD_HEAD, [RequestMethod::METHOD_GET, RequestMethod::METHOD_POST]];
+
+        yield 'OPTIONS: get' => [RequestMethod::METHOD_OPTIONS, [RequestMethod::METHOD_GET]];
+        yield 'OPTIONS: post' => [RequestMethod::METHOD_OPTIONS, [RequestMethod::METHOD_POST]];
+        yield 'OPTIONS: get, post' => [RequestMethod::METHOD_OPTIONS, [RequestMethod::METHOD_GET, RequestMethod::METHOD_POST]];
+    }
+
+    /**
+     * In case we are not using Implicit*Middlewares and we don't have any route with explicit method
+     * returned response should be 405: Method Not Allowed - handled by MethodNotAllowedMiddleware.
+     *
+     * @dataProvider withoutImplicitMiddleware
+     */
+    public function testWithoutImplicitMiddleware(string $method, array $routes)
+    {
+        $router = $this->getRouter();
+        foreach ($routes as $routeMethod) {
+            $route = new Route(
+                '/api/v1/me',
+                $this->prophesize(MiddlewareInterface::class)->reveal(),
+                [$routeMethod]
+            );
+            $router->addRoute($route);
+        }
+
+        $finalResponse = (new Response())->withHeader('mnl-bar', 'foo-baz');
+        $finalResponse->getBody()->write('Final METHOD NOT ALLOWED response.');
```

Comment:
There's really no reason to populate the response. Use a mock here, and test that both the status code and the `Allow` header are set.